scripts/nb-tester/qiskit_docs_notebook_tester/config.py

```diff
@@ -59,159 +59,146 @@ def get_notebook_jobs(args: argparse.Namespace) -> Iterator[NotebookJob]:
             ) from err
 
-    def notebooks_to_execute(self) -> Iterator[Path]:
-        """
-        Yield notebooks to be executed, printing messages for any skipped files.
-        """
-        paths = map(Path, self.args.filenames or self.all_notebooks_to_test)
-        for path in paths:
-            if path.suffix != ".ipynb":
-                print(f"ℹ️ Skipping {path}; file is not `.ipynb` format.")
-                continue
+        groups = [
+            {"name": key, **value} for key, value in config_file["groups"].items()
+        ]
+        test_strategy = args.test_strategy or config_file["default-strategy"]
+        cell_timeout = (
+            args.cell_timeout or
+            config_file.get("test-strategies", {})
+            .get(test_strategy, {})
+            .get("timeout", None)
+        )
+        return cls(
+            cli_filenames=args.filenames,
+            cell_timeout=cell_timeout,
+            test_strategy=test_strategy,
+            write=args.write,
+            groups=groups,
+        )
 
-            if matches(path, self.notebooks_exclude):
-                print(
-                    f"ℹ️ Skipping {path}; to run it, edit `notebooks-exclude` in {self.args.config_path}."
-                )
-                continue
+    def get_patch_for_group(self, group: dict) -> str | None:
+        patch_config = group["test-strategies"].get(self.test_strategy, {})
+        if patch_config == {}:
+            return None
+        return generate_backend_patch(
+            backend_name=patch_config.get("backend", "fake_athens"),
+            provider=patch_config.get("provider", "qiskit-ibm-runtime"),
+            generic_backend_kwargs=patch_config.get("generic_backend_kwargs", None),
+            runtime_service_kwargs=patch_config.get("runtime_service_kwargs", None),
+        )
 
-            if not self.args.submit_jobs and matches(path, self.notebooks_no_mock):
-                print(
-                    f"ℹ️ Skipping {path} as it doesn't work with mock hardware; use the --submit-jobs flag to run it."
-                )
-                continue
 
-            if self.args.only_submit_jobs and not matches(
-                path, self.all_job_submitting_notebooks
-            ):
-                print(
-                    f"ℹ️ Skipping {path} as it doesn't submit jobs and the --only-submit-jobs flag is set."
-                )
-                continue
+def matches(path: Path | str, glob_list: list[str]) -> bool:
+    return any(Path(path).match(glob) for glob in glob_list)
+
+
+def parse_patch_arg(patch: str | None) -> dict:
+    if not patch:
+        return {}
+    try:
+        patch_info = tomllib.loads(f"patch={patch}")["patch"]
+    except tomllib.TOMLDecodeError as err:
+        raise ValueError(
+            "Problem parsing your --patch argument; "
+            "Make sure it's of the form '{ string=\"value\", number=4, ... }'."
+        ) from err
+    return patch_info
```

Comment:
@emilkovacev this is a potentially disruptive change for you, does this new interface meet your needs?

Tl;dr: Rather than having separate CLI arguments for `--provider`/`--backend`/etc., you'd specify the whole patch as a TOML string, e.g.

```
--patch '{ provider="qiskit-fake-provider", backend="fake_athens" }'
```

This patch applies to all filenames passed to the script.

Reply:
I would also propose moving patches into files (like backend.patch), then having an extra required param in the toml for "filename" or "patch_file" that would point to the patch you want to apply. Then you can take the rest of the params in the dict and use `.format()` to inject variables into the patch.

The command would look something like this:
```
--patch '{ file="generic_backend.patch", backend="fake_athens" }'
```

And generic_backend.patch would look like this:
```
from qiskit.providers.fake_provider import GenericBackendV2
def patched_least_busy(self, *args, **kwargs):
    return GenericBackendV2(backend={backend}, ...)
```

In this case, all that would be needed to create a custom patch would be to create a new patch file, no new code is needed.

Reply:
> how would passing multiple patches look?

I hadn't planned to have that functionality through CLI arguments. At the moment, it's one patch for all filepaths passed to the CLI. Do you need the ability to apply different patches to different notebooks without a config file? I'm not sure how that'd look CLI-wise but happy to explore it.

> I would propose moving patches into files...

This is a great idea! I didn't fully understand it when you mentioned it the other day but I do now. I think this could reduce our code complexity a lot and make the interface easier to use.

Reply:
>Do you need the ability to apply different patches to different notebooks without a config file?

Once we introduce more patches for more classes (like SamplerV2 for instance) this would be very useful. 

For instance, if the interface looked like this:
```
--patch '{ file="generic_backend.patch", backend="fake_athens" }' \
--patch '{ file="samplerv2.patch", options={ default_shots=100 } }'
```

Then we could apply layers of patches very easily.

Reply:
@emilkovacev added in 2c25a1844c3787317c38912c63a111e94885468e